Bundle of changes

keks opened the pull request:
This branch bundles a couple of changes so we can test against all of them at once in go-ipfs

- #86 Let cmd.Run return an error (PR #97)
- #98 Remove SetError and add CloseWithError
  - also removes HandleError, RecvdErr and RawNext
  - PRs: #103, #110 
- #102 Change PostRun signature so user don't have to build the chanresponse themselves

keks commented:
@Stebalien 

in one of the PRs above you said you wanted the PR bundled and the changes to go-ipfs ready. This is now done! I'm pretty pleased with the result, especially since it removes some odd stuff like that function `HandleError` or the method `RawNext` on `ResponseEmitter`.

The changes to go-ipfs are in https://github.com/ipfs/go-ipfs/pull/5035.

whyrusleeping commented on cli/responseemitter.go:
```diff
@@ -70,26 +74,52 @@ func (re *responseEmitter) SetEncoder(enc func(io.Writer) cmds.Encoder) {
 }
 
-func (re *responseEmitter) SetError(v interface{}, errType cmdkit.ErrorType) {
+func (re *responseEmitter) CloseWithError(err error) error {
```
It would be really great if this method didnt return an error. I can't think of any situation where the caller wants to do something with the error, or even can. I think it's likely better to just log an error and move on.

keks replied:
The reason I let made this return an error is that `CloseWithError(error) error` is idiomatic and is used throughout the entire Go ecosystem. I don't see a reason to diverge - maybe we don't really need it in ipfs, but possible somewhere else. And now that the command's `Run` function just returns an error we don't call this so often that it's too much of a pain for the caller to do the logging themselves.

keks commented:
Hey @whyrusleeping judging by Stebalien's contribution activity graph it seems they are currently not avalailable, so could you maybe review this PR or delegate this? Thanks!

keks commented:
Also if you have questions on the structure of this bundle PR, shoot!

whyrusleeping commented:
@keks yeah, he's just now getting back from a vacation. I'll still want his review, but i'll take a pass here :)

whyrusleeping commented on chan.go:
```diff
@@ -83,26 +109,33 @@ func (r *chanResponse) Next() (interface{}, error) {
-			r.err = val
-			return nil, ErrRcvdError
+			// TODO keks remove logging
+			log.Error("unexpected error value:", val)
```
rm ?

whyrusleeping commented on chan.go:
```diff
@@ -128,92 +161,94 @@ func (r *chanResponse) RawNext() (interface{}, error) {
+type chanResponseEmitter chanResponse
 
 func (re *chanResponseEmitter) Emit(v interface{}) error {
+	if e, ok := v.(*cmdkit.Error); ok {
```
this is because we shouldnt be emitting errors ever anymore, right?

keks replied:
Exactly. Should I remove this or are we not confident enough this doesn't happen? Or should I return that as an error, so it will definitely fail?

whyrusleeping replied:
i think we should keep it, and maybe even add a check for the non-pointerized version too just in case. Then also add a comment explaining why this bit of code is here.

keks commented:
Hey @Stebalien, do you think you'll find some time to CR this PR and the one in go-ipfs? Would love to merge this soon

whyrusleeping commented on chan.go:
```diff
@@ -70,6 +82,20 @@ func (r *chanResponse) Length() uint64 {
 	return r.length
 }
 
+func (re *chanResponse) Head() Head {
```
Whats this method supposed to do?

keks replied:
It's not new, I just moved it from here: https://github.com/ipfs/go-ipfs-cmds/pull/112/files/15c1508126b5934ffae976e4a09c4353fa90956e#diff-d1319de8df19ba21c67a62fa1fde5f89L201

But I think it's needed for HEAD requests, it's even in the ResponseEmitter interface; but it looks like it's not called from anywhere in `go-ipfs-cmds/...`, so maybe we can drop it. I'd suggest doing that in a new PR though.

keks replied:
I decided to do it in this PR.

Stebalien commented on chan.go:
```diff
@@ -57,7 +61,15 @@ func (r *chanResponse) Error() *cmdkit.Error {
 		return nil
```
Can we get rid of both of these nil tests? I'd rathe panic if something is wrong (and we'll panic in the second case anyways).

keks replied:
well the `r == nil` check obviously doesn't help because we already access `r.wait` before. But what's wrong with the `r.err == nil` check?

keks replied:
We call `r.err.Error()` below, so `r.err` better not be nil.

Stebalien replied:
Sorry, I was refering to both of the `r == nil` checks (in this function and the one above). Unfortunately, git hides that part of the patch.

Stebalien commented on chan.go:
```diff
@@ -70,6 +82,20 @@ func (r *chanResponse) Length() uint64 {
+func (re *chanResponse) Head() Head {
+	<-re.wait
+
+	err, ok := re.err.(*cmdkit.Error)
```
This will treat EOF as an error. Maybe we should just have a separate closed flag if we need that and not write EOF to the error field.

Stebalien replied:
Actually, we can just check `<-re.closed`.

keks replied:
You are right that this is not correct. However, there can be an error even if `re.closed` already is closed, so I don't think it's a good indicator. I think I fixed this locally and will push the changes once I'm through you first pass.

Stebalien commented on chan.go:
```diff
@@ -83,26 +109,29 @@ func (r *chanResponse) Next() (interface{}, error) {
+		return nil, r.err
 	case v, ok := <-r.ch:
 		if !ok {
+			r.ch = nil
```
Are you sure this isn't racy? Also, this'll cause this method to hang forever if called again.

Stebalien commented on chan.go:
```diff
@@ -83,26 +109,29 @@ func (r *chanResponse) Next() (interface{}, error) {
 		}
 
-		if err, ok := v.(cmdkit.Error); ok {
+		if err, ok := v.(*cmdkit.Error); ok {
```
I'm pretty sure we just want to remove this entirely.

keks replied:
I'm not sure, [here](https://github.com/ipfs/go-ipfs-cmds/pull/112#discussion_r192347895) whyrusleeping argued we should check both all the time. There's [an issue](https://github.com/ipfs/go-ipfs-cmdkit/issues/19) on fixing this for `cmdkit.Error` in general, but there was little feedback. 

Also, that line is wrong, should be without the dereference. I'll fix that.

Stebalien replied:
We're already checking if something is an error *before* putting it *into* the channel. If we leave this here, we'll need to leve the `cmdkit.Error` case in the switch statement below. However, IMO, we should just remove both (there shouldn't be *any* errors here).

Stebalien commented on chan.go:
```diff
@@ -128,92 +157,103 @@ func (r *chanResponse) RawNext() (interface{}, error) {
+	}
+	if e, ok := v.(*cmdkit.Error); ok {
+		log.Errorf("unexpected error value emitted: %s at\n%s", e.Error(), debug.Stack())
+	}
```
Shouldn't this call `CloseWithError` and return?

keks replied:
Nah, we don't emit errors anymore. Doing so is a programmer error, so we want them to know. But panicking because of this feels too strong, so I opted for logging an error.

Stebalien replied:
That won't cause a panic, it'll just "do the right thing" (redirect the error). As it stands, we'll end up sending the error on the value channel.

keks replied:
Oh yes, I meant to return. Anyway, I change it to panic now. I hope we can then in the near future remove those panics because we are more confident we don't do that mistake anymore.

Stebalien commented on chan.go:
```diff
@@ -128,92 +157,103 @@ func (r *chanResponse) RawNext() (interface{}, error) {
 
-	close(re.ch)
-	re.ch = nil
+	re.closeOnce.Do(func() {
```
We're calling this under a lock, right? We can just check if it's closed in that case. No need for a `sync.Once`.

keks replied:
Hm, yeah that would also work. I did it this way because IMHO it's more obvious what it's supposed to do than having a `select` reading from `closed` and a `default` case. This one explicitly says "do this once". Also I was under the impression that the stuff in `"sync"` usually is faster than channel ops?

Stebalien replied:
For me, it's more about reducing redundant state. However, benchmarking it, they look about the same (although my benchmark may just suck).

keks replied:
You're right, I got rid of closeOnce.

Stebalien commented on chan.go:
```diff
@@ -128,92 +157,103 @@ func (r *chanResponse) RawNext() (interface{}, error) {
+	})
+
+	// unblock Length(), Error() and Head()
+	re.waitOnce.Do(func() {
```
Same here. If we make sure to always close this under a lock (need to modify `Emit` a bit), we can just use select to see if it's still open.

keks replied:
Again, I don't think doing this with channel operations is more readable or efficient.

Stebalien replied:
For me, every synchronization primitive is something else I have to keep track of. It's easier to keep track of one big lock (and possibly some booleans) than a bunch of `sync.Once`s.

keks replied:
Unfortunately "one big lock" doesn't work in our case (because we need to `Emit` and `Next` at the same time), but I'm down to two and removed all `Once`s.

Stebalien commented on chan.go:
```diff
@@ -128,92 +157,103 @@ func (r *chanResponse) RawNext() (interface{}, error) {
+	re.l.Lock()
+	defer re.l.Unlock()
 
 	if re.ch == nil {
```
I don't think this is a valid "closed" check anymore. Probably want to check:

```go
select {
  case <-re.closed:
    return ...
  default:
}
```

keks replied:
You're right it's not a valid close check anymore, but I think your proposal is also racy: It might occur that re.closed is not yet closed but re.ch is, which would result in a write to closed channel error. Thinking this through...

Stebalien replied:
It should be if we only close the channel while holding the lock.

This is actually a good case for removing the `closeOnce` `sync.Once` (a user may think that `closeOnce` is protecting the close while really the lock is).

Stebalien commented on cli/responseemitter.go:
```diff
@@ -168,29 +210,15 @@ func (re *responseEmitter) Emit(v interface{}) error {
-
-		return nil
-
+		log.Errorf("got an error: %v\n%s", err, debug.Stack())
```
Again, we should probably forward these errors to `CloseWithError` for now.

keks replied:
I believe we should do away with emitting errors and make sure doing so is an error.

Stebalien replied:
We should. However, we should either *panic* in cases like this or "do the right thing". Simply logging and dropping the error on the floor is not going to end well.

Stebalien commented on cli/responseemitter.go:
```diff
@@ -168,29 +210,15 @@ func (re *responseEmitter) Emit(v interface{}) error {
 		if err != nil {
-			re.SetError(err, cmdkit.ErrNormal)
-			err = nil
+			return err
```
I'm wondering if this should also call `CloseWithError`. We could also say that it's the callers responsibility to return the error from `Run`. Thoughts?

keks replied:
Calling `CloseWithError` here would close the emitter and only print `io.Copy`'s error. I think it's better to allow the caller to specify a more intelligible error.

Stebalien replied:
Fair enough.

Stebalien commented on cli/responseemitter_test.go:
```diff
@@ -2,10 +2,11 @@ package cli
 	"testing"
 
-	"github.com/ipfs/go-ipfs-cmdkit"
+	//"github.com/ipfs/go-ipfs-cmdkit"
```
Should probably remove before final merge.

keks replied:
yup.

Stebalien commented on command.go:
```diff
@@ -68,24 +68,39 @@ var ErrIncorrectType = errors.New("The command returned a value with a different
-	defer re.Close()
+	var err error
+
+	defer func() {
```
How about putting all the logic in `Call` in a separate `call` function that turns an error. That way, we don't need this defer? The shared `err` variable makes be uncomfortable.

Stebalien commented on http/response.go:
```diff
@@ -67,13 +70,29 @@ func (res *Response) RawNext() (interface{}, error) {
+	return v, err
 }
 
 func (res *Response) Next() (interface{}, error) {
```
THis function checks if the value is an error, I assume this isn't possible given RawNext?

Stebalien commented on http/responseemitter.go:
```diff
@@ -147,12 +160,28 @@ func (re *responseEmitter) Close() error {
+		e = &cmdkit.Error{Message: err.Error()}
+	}
+
+	if re.streaming || e.Code == cmdkit.ErrFatal {
```
Aren't all errors "fatal" now? That is, can't we just *always* use the trailer now?

Stebalien commented on writer.go:
```diff
@@ -94,11 +95,10 @@ func (r *readerResponse) Next() (interface{}, error) {
-		return nil, ErrRcvdError
 	case Single:
 		return val.Value, nil
+	case *cmdkit.Error:
```
Again, I assume this case is impossible.

Stebalien commented:
Nice! This is starting to look a lot more robust.

keks commented:
@Stebalien thank you for the thorough feedback! I tried to incorporate all of it and made some additional changes:
- remove the `Header` interface and the `Head()` methods from all `Response` types. It wasn't being used anywhere.
- remove the `RawNext` methods from all `Response` types
- remove the `HandleError` function
- unexport `WriterResponseEmitter`
- add a new global error variable `ErrClosedEmitter` that is meant to be returned by `Emit` and `Close{,WithError}`.
- work a little bit on testing. More to come in a different PR.

Right now Travis complains that `writer.go` is not gofmt'd. However, when doing `go fmt ./...`, no file is changed. I'm not sure what's going on here.

Also, in my branch of go-ipfs that has the changes for this library there is a test ([this one]) that doesn't complete currently. Investigating that now. Also my branch fell behind a bit, so I'll still need to rebase that on master -- but I first want to get that test to pass.

[this one]: https://github.com/ipfs/go-ipfs/blob/master/test/sharness/t0600-issues-and-regressions-online.sh#L49-L51

keks commented:
Okay, I fixed all the tests in the go-ipfs branch and rebased it is it's en par with master.

I still have no clue with Travis let's this PR fail though. It thinks that writer.go is not gofmt'd, even though it is -- at least for me calling go fmt doesn't change anything. Rebuilding doesn't change anything.
Has anyone of you stumbled over something like this in the past? My go installation is up to date.

keks commented:
cc @Stebalien

Stebalien commented on http/responseemitter.go:
```diff
@@ -67,59 +69,57 @@ type responseEmitter struct {
-		ch, isChan = value.(chan interface{})
+	if single, ok := value.(cmds.Single); ok {
+		value = single.Value
+		defer re.closeWithError(nil)
```
Probably needs to go after the lock.

Stebalien commented on http/responseemitter.go:
```diff
@@ -67,59 +69,57 @@ type responseEmitter struct {
-	if _, ok := value.(cmdkit.Error); ok {
-		log.Warning("fixme: got Error not *Error: ", value)
-		value = &value
+	if err, ok := value.(cmdkit.Error); ok {
```
Use the debug error assertion?

Stebalien commented on writer.go:
```diff
@@ -117,78 +101,88 @@ type WriterResponseEmitter struct {
+}
+
+func (re *writerResponseEmitter) SetError(v interface{}, errType cmdkit.ErrorType) {
 	err := re.Emit(&cmdkit.Error{Message: fmt.Sprint(v), Code: errType})
```
Shouldn't we remove this function?

keks replied:
We can't just remove it, because then it wouldn't satisfy the interface anymore. But I made it simply log an error.

keks commented on executor.go:
```diff
@@ -105,6 +132,12 @@ func (x *executor) Execute(req *Request, re ResponseEmitter, env Environment) (e
 		}
 
```
I'd still want to return the error from `errCh`, but yes the defer should also wait for it. So let's wait in both places.

keks commented on executor.go:
```diff
@@ -77,26 +76,54 @@ func (x *executor) Execute(req *Request, re ResponseEmitter, env Environment) (e
+
 			// if they are errors
-			if e, ok := v.(error); ok {
+			if err, ok := v.(error); ok {
```
I know it's creating a new error variable, and I use it as the argument when calling `CloseWithError`. I don't think I'm throwing it away?

keks commented on chan.go:
```diff
@@ -2,71 +2,83 @@ package cmds
 }
 
 func (r *chanResponse) Error() *cmdkit.Error {
 	<-r.wait
```
I understand the concern about deadlocks, but returning immediately raises the question of what to return when the command is still running and we didn't see an error yet. So I went for wait-until-close, I'm currently testing this in go-ipfs to see if it makes problems.

keks replied:
Okay, if we block on error until the stream is closed and change nothing else, we get a deadlock. Now I just return nil if the connection is not closed yet.

Stebalien commented on writer.go:
```diff
@@ -117,78 +101,86 @@ type WriterResponseEmitter struct {
+	debug.AssertNotError(v)
 
 	if _, ok := v.(Single); ok {
 		defer re.Close()
```
Will close without error if encoding fails.

keks replied:
I don't think I would send the encoding error down the stream. If I try to send an unencodable value through a writer, the stream should live on and I should get an error so I can send something else.

However, we should return the `Close()` error, if there is one.

Stebalien commented on writer.go:
```diff
@@ -117,78 +101,86 @@ type WriterResponseEmitter struct {
-func (re *WriterResponseEmitter) SetLength(length uint64) {
+func (re *writerResponseEmitter) SetError(v interface{}, errType cmdkit.ErrorType) {
+	err := &cmdkit.Error{Message: fmt.Sprint(v), Code: errType}
+	log.Errorf("error setting error %q: cannot set error with writerResponseEmitter", err)
```
If we can't do the right thing here, maybe we should just remove this from the interface. We really don't want the user to try calling this and then return a "success" to the user.

Stebalien replied:
Actually, *hasn't* this been removed from the interface?

keks replied:
Yes, looks like I forgot deleting it. done.

Stebalien commented on http/responseemitter.go:
```diff
@@ -67,59 +69,50 @@ type responseEmitter struct {
+	
+	if single, ok := value.(cmds.Single); ok {
+		value = single.Value
+		defer re.closeWithError(nil)
```
Will throw away encoding errors.

keks replied:
IMHO encoding errors should be passed to the caller instead of be used to close the stream. And `closeWithError` always returns `nil` so I figured we don't need it to handle the error.

Stebalien replied:
The issue is that we're closing the response stream as if the command succeeded when, really, the command failed.

Stebalien commented:
FYI, if I take more than a week to get to a review, the notification has been buried in my inbox. At that point, just bug me on IRC (really, feel free to bug me on IRC repeatedly as soon as you have something for me to review).

keks commented:
Thank you for the review. I was traveling and will work on these over the next week. And okay, next time I'll be more stubborn :)

keks commented on chan.go:
```diff
@@ -85,135 +103,121 @@ func (r *chanResponse) Next() (interface{}, error) {
 			return v, nil
 		}
 	case <-ctx.Done():
-		close(r.done)
```
I'm not sure. That would mean that if one call to `Next()` gets cancelled, the stream will end. Intuitively I would say that the stream should live on and the caller should be able to do another call to `Next()` later. Don't you think?

Stebalien replied:
Ah, no, you're totally right.

keks commented:
Hey @Stebalien I now commented or acted on everything except the thoughts on `Single`. I'm still pondering on that one, especially whether it's possible to only set the headers that this is a single value, but still leave the closing to the caller. Otherwise making `Close()` idempotent, i.e. closing a closed channel doesn't do anything, also seems like a good idea. Giving that some thought right now.

keks commented:
I think the most idiomatic way is to ignore double close errors when getting an error on `Close` or `CloseWithError`, at least in places where we expect these to occur.
For this I'd introduce a new error variable `ErrEmitterClosed` and return that on double close. On the caller side I'd always check errors like
```go
err = re.Close()
if err != ErrEmitterClosed && err != nil {
  // handle error
}
```

Stebalien commented:
>  ignore this error in 

* Call
* Execute

And possibly some places in go-ipfs.

---

Warning, @whyrusleeping is grumbling about this being a massive breaking change. My plan is:

1. Finish up this PR but don't merge it:
2. *Try* patching go-ipfs.
3. If the patch is too hard to write or is unreviewable:
    1. Make the `Run` function an `interface{}`.
    2. Type switch on error/no error.
    3. Add back a `SetError` function that just calls `CloseWithError` under the hood.

Stebalien commented on http/parse.go:
```diff
@@ -211,10 +212,10 @@ func parseResponse(httpRes *http.Response, req *cmds.Request) (cmds.Response, er
-			if err != nil {
-				return nil, err
+			// handle errors from headers
+			errStr := httpRes.Header.Get(StreamErrHeader)
```
Is this correct? I think we want to set `res.initErr` and only *return* an error if we have some kind of parse error, right?

---

Also, we should probably maintain compatibility with existing servers (for now). We should probably.

1. Check the header first.
2. Check the stream next.

Speaking of which, we may want to do the inverse: Always send errors as values along with the header (just in case). Is that possible?

---

Basically, I'm worried about the case where a user calls `ipfs --api=/some/remote/server ...`.

keks replied:
> Is this correct? I think we want to set res.initErr and only return an error if we have some kind of parse error, right?

Well, `parseResponse` is called by `Send` which is called by `Execute`, so the difference is whether you receive the error from `Send`/`Execute` or from `Next()`. IMHO returning the error from `Send` is fine, and making `Next` return the error would make the code more complex because you would have to make a noop `Response` that only returns an error and make `parseResponse` return that. I mean we can totally do that but I don't see the benefit.

> Also, we should probably maintain compatibility with existing servers (for now). We should probably...

IIUC the header is only set on connection close so I'd only check it when we receive an EOF. Does that sound reasonable to you?

> Speaking of which, we may want to do the inverse: Always send errors as values along with the header (just in case). Is that possible?

I think since the other side always checks both, it's okay to only set the header, right?

> Basically, I'm worried about the case where a user calls ipfs --api=/some/remote/server ....

Hmm, I don't get what's so special about that one. Can you elaborate?

keks replied:
Oh I see what you mean now. Yes setting `res.initErr` seems to be the right thing to do.

Stebalien replied:
> IIUC the header is only set on connection close so I'd only check it when we receive an EOF. Does that sound reasonable to you?

Yes.

> I think since the other side always checks both, it's okay to only set the header, right?

Is that *currently* true?

> Hmm, I don't get what's so special about that one. Can you elaborate?

The client version may not match the server version. If the current client doesn't always read the header or the current server version doesn't always send it, bad things may happen.

keks replied:
Hmm, I'm afraid you are right and we'll have to send errors also as values. Damn, I hoped we would not have to do that anymore...

Stebalien commented on http/response.go:
```diff
@@ -74,34 +85,27 @@ func (res *Response) RawNext() (interface{}, error) {
 	if err != nil {
-		return nil, err
+		if err == io.EOF {
+			res.err = err
```
I may be missing something, but do we need to check for trailing error headers here?

keks replied:
Yes, you're totally right. Building a comprehensive test suite for cmds would be a big and important project by itself...

Stebalien commented on http/responseemitter.go:
```diff
@@ -136,23 +129,43 @@ func (re *responseEmitter) Emit(value interface{}) error {
-		log.Debug("http.SetError err=", err)
-		panic(err)
+		// abort by sending an error trailer
+		re.w.Header().Set(StreamErrHeader, err.Error())
```
This is where we'd *also* send an error value if we decided to go that route. Seeing this, I'm now pretty sure we're not reading an error somewhere where we should be (I don't see us *reading* this trailer anywhere).

keks replied:
I now check the header errors here: https://github.com/ipfs/go-ipfs-cmds/pull/112/files/#diff-c8a8dcac9ba61ed253376bf559309ae4R90

keks commented on chan.go:
```diff
@@ -85,135 +102,121 @@ func (r *chanResponse) Next() (interface{}, error) {
-	re.ch = nil
 
-	return nil
+	// unblock Length() and Error()
```
I was referring to both selects below that comment, but to clarify I'll just make two comments instead.

Stebalien replied:
As far as I can tell, `Error()` doesn't block on anything.

keks commented on cli/responseemitter.go:
```diff
@@ -70,26 +71,50 @@ func (re *responseEmitter) SetEncoder(enc func(io.Writer) cmds.Encoder) {
+	_, err = fmt.Fprintln(re.stderr, "Error:", e.Message)
 	if err != nil {
-		panic(err)
+		return err
```
We check the errors in `Emit`, so I'd keep it in there. Also the user can specify which readers and writers to use for stdout and stderr and these don't have to be `os.Stdout` and `os.Stderr`, so it's difficult to make assumptions here.

Stebalien replied:
Got it. Fair enough.

keks commented:
Yup this PR is pretty big, but I have a [go-ipfs branch](https://github.com/ipfs/go-ipfs/tree/feat/cmds2/) that worked with this branch before your CR. It has fallen behind by now, but we didn't really change anything big on this side. I expect the main issue is that I'll have to rebase it to go-ipfs master.

keks commented:
Here is the accompanying PR: https://github.com/ipfs/go-ipfs/pull/5035

Stebalien commented:
> Here is the accompanying PR: ipfs/go-ipfs#5035

Nice! That's totally reviewable.

Stebalien commented on chan_test.go:
```diff
@@ -78,43 +78,6 @@ func TestChanResponsePair(t *testing.T) {
 	}
 }
 
-func TestSingle1(t *testing.T) {
```
What was racy about this test?

keks replied:
I'm not 100% sure myself. The race detector complains that the [`Close` call](https://github.com/ipfs/go-ipfs-cmds/pull/112/files/1428e703db2cf9135d0c1f710fac9391f48716d1..a011832acc5d8d1bd013da9d1422f9f497989ede#diff-28e0ba9683655894160fb3fe3c0c6aceL94) and the [`Next` call](https://github.com/ipfs/go-ipfs-cmds/pull/112/files/1428e703db2cf9135d0c1f710fac9391f48716d1..a011832acc5d8d1bd013da9d1422f9f497989ede#diff-28e0ba9683655894160fb3fe3c0c6aceL101) are unsynchronized writes, but I just don't see it. The test was basically the same as `TestSingleChan`, so I removed this one.

Stebalien replied:
Then there is a bug somewhere chanResponse.

Stebalien replied:
Ah. I see, the test is racy.

Stebalien commented on cli/single_test.go:
```diff
@@ -0,0 +1,54 @@
+			t.Fatal(err)
+		}
+
+		wait <- struct{}{}
```
Why do we need to wait here?

keks replied:
That line is supposed to unblock the wait outside the goroutine. I added that before I read the `exitCh`, but yes we don't need that anymore.

Stebalien commented on command.go:
```diff
@@ -113,9 +114,7 @@ func (c *Command) call(req *Request, re ResponseEmitter, env Environment) error
-			// TODO or should this return an error?
-			log.Errorf("unknown encoding %q, using json", encType)
-			re_.SetEncoder(Encoders[JSON](req))
+			return fmt.Errorf("unknown encoding %q, using json", encType)
```
Not using json.

Stebalien commented on http/response.go:
```diff
@@ -87,6 +87,12 @@ func (res *Response) Next() (interface{}, error) {
+			// handle errors from headers
+			errStr := res.res.Header.Get(StreamErrHeader)
+			if errStr != "" {
+				return nil, &cmdkit.Error{Message: errStr}
```
Still needs to *set* the error.

Stebalien commented on http/responseemitter.go:
```diff
@@ -146,6 +143,12 @@ func (re *responseEmitter) CloseWithError(err error) error {
+		return cmds.ErrClosingClosedEmitter
+	}
+
+	re.closed = true
```
Shouldn't need to do this twice. We already do this in `closeWithError`.

Stebalien commented on chan_test.go:
```diff
@@ -78,43 +78,6 @@ func TestChanResponsePair(t *testing.T) {
-	go func() {
-		re.Emit(Single{42})
-
-		err = re.Close()
```
This should be `:=`

Stebalien replied:
Fixed.

keks replied:
Doh, of course! Thanks.

Stebalien commented on http/parse.go:
```diff
@@ -190,7 +190,11 @@ func parseResponse(httpRes *http.Response, req *cmds.Request) (cmds.Response, er
 		if ok {
 			res.dec = makeDec(res.rr)
+		} else {
+			log.Errorf("could not find decoder for encoding %q", encType)
```
Not sure if this is correct. It'll probably warn on plain text.

Stebalien commented on http/responseemitter.go:
```diff
@@ -212,8 +227,10 @@ func (re *responseEmitter) doPreamble(value interface{}) {
 		} else {
 			status = http.StatusInternalServerError
 		}
+		h.Set(StreamErrHeader, err.Message)
```
This will set a *header* but not a *trailer*.

Stebalien replied:
That is, it'll only work if we haven't sent a value.

keks replied:
Oh, apparently I deleted actually setting the trailer in `closeWithError`. I added it again, and now I always set it on non-nil errors.
Should I change that to only set the headers when we send a byte stream/`io.Reader`?

keks commented:
@Stebalien FYI, I rebased the branches in go-ipfs-cmds and go-ipfs and updated cmds' version in go-ipfs. This made some of the sharness tests unhappy and I'm working on getting those passing again.

Stebalien commented:
So, it turns out I was *wrong* about how we currently handle HTTP headers/errors. We *always* handle trailing errors using the `responseReader` helper (so we don't need to do that twice). My comment was incorrect: https://github.com/ipfs/go-ipfs-cmds/pull/112#discussion_r209087640

Given this, I believe we can switch to sending errors-in-trailers *only* for all 200 responses (don't *additionally* include the value). Thoughts? This should cover both the old client talking to a new server and a new server talking to an old client cases.

I'd like to eventually move to:

1. Sending a non-200 response on parse failure or command not found (returning the error as a plain string in the response).
2. Sending a 200 response in all other cases, returning any errors in a trailer.

But we can tackle this later.

keks commented:
@Stebalien done! Want to give it another round of review?

Stebalien commented on http/parse.go:
```diff
@@ -218,7 +218,10 @@ func parseResponse(httpRes *http.Response, req *cmds.Request) (cmds.Response, er
 			return nil, fmt.Errorf("unknown error content type: %s", contentType)
 		default:
 			// handle errors from headers
-			e.Message = httpRes.Header.Get(StreamErrHeader)
```
Should we not be *returning* an error in this case?

keks replied:
As it stands now the error will be returned on the first call to `Next()`, but I can also make it return immediately. I believe that made the test case more complex, but maybe that's okay.

Stebalien commented on http/responseemitter.go:
```diff
@@ -112,19 +111,23 @@ func (re *responseEmitter) Emit(value interface{}) error {
-		err = re.closeWithError(err)
+		// always close singles with nil error.
+		// encoding errors go to caller.
+		err = re.closeWithError(nil)
```
If we return the error to the "caller", expecting them to do something, we *can't* close. This should be something like `isSingle && err != nil`. Am I missing something?

keks replied:
It should have been `isSingle && err == nil`. If an encoding or copy error occurs, we didn't send the value so it doesn't make sense to close the emitter but we should return it. If encoding went fine, gracefully close the emitter.

Stebalien commented:
You know, I think this is pretty much ready (modulo those two questions/issues). This is going to be awesome!

keks commented:
Hey @Stebalien, please CR :)

Stebalien commented:
@keks all ready to go! I'll merge this as soon as https://github.com/ipfs/go-ipfs/pull/5035 is ready.
